packages/lib/zuauth/src/server.ts

```diff
@@ -34,37 +99,66 @@ export async function authenticate(
     throw new Error("ZK ticket PCD is not valid");
   }
 
-  if (pcd.claim.watermark.toString() !== watermark) {
-    throw new Error("PCD watermark doesn't match");
+  /**
+   * The configuration array must not be empty.
+   */
+  if (config.length === 0) {
+    throw new Error("Configuration is empty");
   }
 
-  const publicKeys = config.map((em) => em.publicKey);
-  const productIds = new Set(
-    // Product ID is optional, so it's important to filter out undefined values
-    config
-      .map((em) => em.productId)
-      .filter((productId) => productId !== undefined)
-  );
-
-  if (
-    publicKeys.length > 0 &&
-    !publicKeys.find((pubKey) =>
-      isEqualEdDSAPublicKey(pubKey, pcd.claim.signer)
-    )
-  ) {
+  /**
+   * Check if the external nullifier matches the configuration.
+   */
+  if (externalNullifier !== undefined) {
+    if (pcd.claim.externalNullifier === undefined) {
+      throw new Error(
+        "PCD is missing external nullifier when one was provided"
+      );
+    }
+    if (
+      pcd.claim.externalNullifier.toString() !== externalNullifier.toString()
+    ) {
+      throw new Error("External nullifier does not match the provided value");
+    }
+  } else if (pcd.claim.externalNullifier !== undefined) {
     throw new Error(
-      "Signing key does not match any of the configured public keys"
+      "PCD contains an external nullifier when none was provided"
     );
   }
 
-  if (
-    productIds.size > 0 &&
-    pcd.claim.partialTicket.productId &&
-    !productIds.has(pcd.claim.partialTicket.productId)
-  ) {
-    throw new Error(
-      "Product ID does not match any of the configured product IDs"
-    );
+  if (pcd.claim.watermark !== watermark.toString()) {
+    throw new Error("PCD watermark does not match");
+  }
+
+  /**
+   * Check that the revealed fields in the PCD match the expectations set out
+   * in {@link revealedFields}. This is to ensure the consistency between the
+   * configuration passed to this function, and the configuration used on the
+   * client-side when generating the PCD.
+   */
+  for (const [revealedField, fieldName] of Object.entries(revealedFields)) {
+    if (fieldsToReveal[revealedField as keyof EdDSATicketFieldsToReveal]) {
+      checkIsDefined(pcd.claim.partialTicket[fieldName], fieldName);
+    } else {
+      checkIsUndefined(pcd.claim.partialTicket[fieldName], fieldName);
+    }
+  }
+
```

Comment:
If you're now desupporting validEventIds, should you check here that it's required to be undefined?  I don't think it changes the security (given you're checking that all the necessary fields are being revealed), but it would catch a configuration error.

Reply:
I've added a check for this too.